File: src-tauri/src/commands/provider.rs
use crate::models::result::CommandResult;
use crate::services::provider_service::{
    self, CreateProviderPayload, DeleteProviderData, Provider, ProviderValidationData,
    UpdateProviderPayload,
};

#[tauri::command]
pub async fn list_providers() -> CommandResult<Vec<Provider>> {
    match provider_service::list_providers().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn create_provider(payload: CreateProviderPayload) -> CommandResult<Provider> {
    match provider_service::create_provider(payload).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn update_provider(payload: UpdateProviderPayload) -> CommandResult<Provider> {
    match provider_service::update_provider(payload).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn delete_provider(id: String) -> CommandResult<DeleteProviderData> {
    match provider_service::delete_provider(id).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn validate_provider(id: String) -> CommandResult<ProviderValidationData> {
    match provider_service::validate_provider(id).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

File: src/main/ipc/provider.ts
import { ipcMain } from "electron";
import { IPC } from "../../shared/ipc-channels.js";
import { ok, err } from "../models/result.js";
import { toAppError } from "./utils.js";
import {
  listProviders,
  createProvider,
  updateProvider,
  deleteProvider,
  validateProvider,
} from "../services/provider-service.js";
import type {
  CreateProviderPayload,
  UpdateProviderPayload,
} from "../services/provider-service.js";

export function registerProviderHandlers(): void {
  ipcMain.handle(IPC.LIST_PROVIDERS, async () => {
    try { return ok(await listProviders()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.CREATE_PROVIDER, async (_event, payload: CreateProviderPayload) => {
    try { return ok(await createProvider(payload)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.UPDATE_PROVIDER, async (_event, payload: UpdateProviderPayload) => {
    try { return ok(await updateProvider(payload)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.DELETE_PROVIDER, async (_event, payload: { id: string }) => {
    try { return ok(await deleteProvider(payload.id)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.VALIDATE_PROVIDER, async (_event, payload: { id: string }) => {
    try { return ok(await validateProvider(payload.id)); }
    catch (error) { return err(toAppError(error)); }
  });
}
